syntax = "proto3";

option optimize_for = LITE_RUNTIME;

service ProtoHailoRtRpc {
    rpc client_keep_alive (keepalive_Request) returns (empty) {}
    rpc get_service_version (get_service_version_Request) returns (get_service_version_Reply) {}

    rpc VDevice_create (VDevice_create_Request) returns (VDevice_create_Reply) {}
    rpc VDevice_dup_handle (dup_handle_Request) returns (dup_handle_Reply) {}
    rpc VDevice_release (Release_Request) returns (Release_Reply) {}
    rpc VDevice_configure (VDevice_configure_Request) returns (VDevice_configure_Reply) {}
    rpc VDevice_get_physical_devices_ids (VDevice_get_physical_devices_ids_Request) returns (VDevice_get_physical_devices_ids_Reply) {}
    rpc VDevice_get_default_streams_interface (VDevice_get_default_streams_interface_Request) returns (VDevice_get_default_streams_interface_Reply) {}

    rpc ConfiguredNetworkGroup_dup_handle (dup_handle_Request) returns (dup_handle_Reply) {}
    rpc ConfiguredNetworkGroup_release (Release_Request) returns (Release_Reply) {}
    rpc ConfiguredNetworkGroup_make_input_vstream_params (ConfiguredNetworkGroup_make_input_vstream_params_Request) returns (ConfiguredNetworkGroup_make_input_vstream_params_Reply) {}
    rpc ConfiguredNetworkGroup_make_output_vstream_params (ConfiguredNetworkGroup_make_output_vstream_params_Request) returns (ConfiguredNetworkGroup_make_output_vstream_params_Reply) {}
    rpc ConfiguredNetworkGroup_make_output_vstream_params_groups (ConfiguredNetworkGroup_make_output_vstream_params_groups_Request) returns (ConfiguredNetworkGroup_make_output_vstream_params_groups_Reply) {}
    rpc ConfiguredNetworkGroup_name (ConfiguredNetworkGroup_name_Request) returns (ConfiguredNetworkGroup_name_Reply) {}
    rpc ConfiguredNetworkGroup_get_network_infos (ConfiguredNetworkGroup_get_network_infos_Request) returns (ConfiguredNetworkGroup_get_network_infos_Reply) {}
    rpc ConfiguredNetworkGroup_get_all_stream_infos (ConfiguredNetworkGroup_get_all_stream_infos_Request) returns (ConfiguredNetworkGroup_get_all_stream_infos_Reply) {}
    rpc ConfiguredNetworkGroup_get_default_stream_interface (ConfiguredNetworkGroup_get_default_stream_interface_Request) returns (ConfiguredNetworkGroup_get_default_stream_interface_Reply) {}
    rpc ConfiguredNetworkGroup_get_output_vstream_groups (ConfiguredNetworkGroup_get_output_vstream_groups_Request) returns (ConfiguredNetworkGroup_get_output_vstream_groups_Reply) {}
    rpc ConfiguredNetworkGroup_get_input_vstream_infos (ConfiguredNetworkGroup_get_vstream_infos_Request) returns (ConfiguredNetworkGroup_get_vstream_infos_Reply) {}
    rpc ConfiguredNetworkGroup_get_output_vstream_infos (ConfiguredNetworkGroup_get_vstream_infos_Request) returns (ConfiguredNetworkGroup_get_vstream_infos_Reply) {}
    rpc ConfiguredNetworkGroup_get_all_vstream_infos (ConfiguredNetworkGroup_get_vstream_infos_Request) returns (ConfiguredNetworkGroup_get_vstream_infos_Reply) {}
    rpc ConfiguredNetworkGroup_is_scheduled (ConfiguredNetworkGroup_is_scheduled_Request) returns (ConfiguredNetworkGroup_is_scheduled_Reply) {}
    rpc ConfiguredNetworkGroup_set_scheduler_timeout (ConfiguredNetworkGroup_set_scheduler_timeout_Request) returns (ConfiguredNetworkGroup_set_scheduler_timeout_Reply) {}
    rpc ConfiguredNetworkGroup_set_scheduler_threshold (ConfiguredNetworkGroup_set_scheduler_threshold_Request) returns (ConfiguredNetworkGroup_set_scheduler_threshold_Reply) {}
    rpc ConfiguredNetworkGroup_set_scheduler_priority (ConfiguredNetworkGroup_set_scheduler_priority_Request) returns (ConfiguredNetworkGroup_set_scheduler_priority_Reply) {}
    rpc ConfiguredNetworkGroup_get_latency_measurement (ConfiguredNetworkGroup_get_latency_measurement_Request) returns (ConfiguredNetworkGroup_get_latency_measurement_Reply) {}
    rpc ConfiguredNetworkGroup_is_multi_context (ConfiguredNetworkGroup_is_multi_context_Request) returns (ConfiguredNetworkGroup_is_multi_context_Reply) {}
    rpc ConfiguredNetworkGroup_get_config_params(ConfiguredNetworkGroup_get_config_params_Request) returns (ConfiguredNetworkGroup_get_config_params_Reply) {}

    rpc InputVStreams_create (VStream_create_Request) returns (VStreams_create_Reply) {}
    rpc InputVStream_dup_handle (dup_handle_Request) returns (dup_handle_Reply) {}
    rpc OutputVStream_dup_handle (dup_handle_Request) returns (dup_handle_Reply) {}
    rpc InputVStream_release (Release_Request) returns (Release_Reply) {}
    rpc OutputVStreams_create (VStream_create_Request) returns (VStreams_create_Reply) {}
    rpc OutputVStream_release (Release_Request) returns (Release_Reply) {}
    rpc InputVStream_write (InputVStream_write_Request) returns (InputVStream_write_Reply) {}
    rpc OutputVStream_read (OutputVStream_read_Request) returns (OutputVStream_read_Reply) {}
    rpc InputVStream_get_frame_size (VStream_get_frame_size_Request) returns (VStream_get_frame_size_Reply) {}
    rpc OutputVStream_get_frame_size (VStream_get_frame_size_Request) returns (VStream_get_frame_size_Reply) {}
    rpc InputVStream_flush (InputVStream_flush_Request) returns (InputVStream_flush_Reply) {}
    rpc InputVStream_name (VStream_name_Request) returns (VStream_name_Reply) {}
    rpc OutputVStream_name (VStream_name_Request) returns (VStream_name_Reply) {}
    rpc InputVStream_network_name (VStream_network_name_Request) returns (VStream_network_name_Reply) {}
    rpc OutputVStream_network_name (VStream_network_name_Request) returns (VStream_network_name_Reply) {}
    rpc InputVStream_abort (VStream_abort_Request) returns (VStream_abort_Reply) {}
    rpc OutputVStream_abort (VStream_abort_Request) returns (VStream_abort_Reply) {}
    rpc InputVStream_resume (VStream_resume_Request) returns (VStream_resume_Reply) {}
    rpc OutputVStream_resume (VStream_resume_Request) returns (VStream_resume_Reply) {}
    rpc InputVStream_get_user_buffer_format (VStream_get_user_buffer_format_Request) returns (VStream_get_user_buffer_format_Reply) {}
    rpc OutputVStream_get_user_buffer_format (VStream_get_user_buffer_format_Request) returns (VStream_get_user_buffer_format_Reply) {}
    rpc InputVStream_get_info (VStream_get_info_Request) returns (VStream_get_info_Reply) {}
    rpc OutputVStream_get_info (VStream_get_info_Request) returns (VStream_get_info_Reply) {}
}

message empty {}

message keepalive_Request {
    uint32 pid = 1;
}

message ProtoVDeviceParams {
    uint32 device_count = 1;
    repeated string device_ids = 2;
    uint32 scheduling_algorithm = 3;
    string group_id = 4;
}

message ProtoHailoVersion {
    uint32 major_version = 1;
    uint32 minor_version = 2;
    uint32 revision_version = 3;
}

message get_service_version_Request {
}

message get_service_version_Reply {
    uint32 status = 1;
    ProtoHailoVersion hailo_version = 2;
}

message dup_handle_Request {
    uint32 pid = 1;
    uint32 handle = 2;
}

message dup_handle_Reply {
    uint32 handle = 1;
}

message VDevice_create_Request {
    ProtoVDeviceParams hailo_vdevice_params = 1;
    uint32 pid = 2;
}

message VDevice_create_Reply {
    uint32 status = 1;
    uint32 handle = 2;
}

message Release_Request {
    uint32 handle = 1;
}

message Release_Reply {
    uint32 status = 1;
}

message VStreams_create_Reply {
    uint32 status = 1;
    repeated uint32 handles = 2;
}

message VStream_create_Request {
    uint32 net_group = 1;
    repeated ProtoNamedVStreamParams vstreams_params = 2;
    uint32 pid = 3;
}

message ProtoHailoFormat {
    uint32 type = 1;
    uint32 order = 2;
    uint32 flags = 3;
}

message ProtoVStreamParams {
    ProtoHailoFormat user_buffer_format = 1;
    uint32 timeout_ms = 2;
    uint32 queue_size = 3;
    uint32 vstream_stats_flags = 4;
    uint32 pipeline_elements_stats_flags = 5;
}

message ProtoNamedVStreamParams {
    string name = 1;
    ProtoVStreamParams params = 2;
}

message ProtoThreeDImageShape {
    uint32 height = 1;
    uint32 width = 2;
    uint32 features = 3;
}

message ProtoStreamShape {
    ProtoThreeDImageShape shape = 1;
    ProtoThreeDImageShape hw_shape = 2;
}

message ProtoNmsDefuseInfo {
    uint32 class_group_index = 1;
    string original_name = 2;
}

message ProtoNmsInfo {
    uint32 number_of_classes = 1;
    uint32 max_bboxes_per_class = 2;
    uint32 bbox_size = 3;
    uint32 chunks_per_frame = 4;
    bool is_defused = 5;
    ProtoNmsDefuseInfo defuse_info = 6;
}

message ProtoQuantInfo {
    float qp_zp = 1;
    float qp_scale = 2;
    float limvals_min = 3;
    float limvals_max = 4;
}

message ProtoStreamInfo {
    ProtoStreamShape stream_shape = 1;
    ProtoNmsInfo nms_info = 2;
    uint32 hw_data_bytes = 3;
    uint32 hw_frame_size = 4;
    ProtoHailoFormat format = 5;
    uint32 direction = 6;
    uint32 index = 7;
    string name = 8;
    ProtoQuantInfo quant_info = 9;
    bool is_mux = 10; 
}

message ProtoStreamsParams {
    uint32 stream_interface = 1;
    uint32 direction = 2;
    uint32 flags = 3;
}

message ProtoNamedStreamParams {
    string name = 1;
    ProtoStreamsParams params = 2;
}

message ProtoNetworkParams {
    uint32 batch_size = 1;
}

message ProtoNamedNetworkParams {
    string name = 1;
    ProtoNetworkParams params = 2;
}

message ProtoNmsShape {
    uint32 number_of_classes = 1;
    uint32 max_bbox_per_class = 2;
}

message ProtoVStreamInfo {
    string name = 1;
    string network_name = 2;
    uint32 direction = 3;
    ProtoHailoFormat format = 4;
    ProtoThreeDImageShape shape = 5;
    ProtoNmsShape nms_shape = 6;
    ProtoQuantInfo quant_info = 7;
}

message ProtoConfigureNetworkParams {
    uint32 batch_size = 1;
    uint32 power_mode = 2;
    uint32 latency = 3;
    repeated ProtoNamedStreamParams stream_params_map = 4;
    repeated ProtoNamedNetworkParams network_params_map = 5;
}

message ProtoNamedConfigureNetworkParams {
    string name = 1;
    ProtoConfigureNetworkParams params = 2;
}

message ProtoVStreamGroup {
    repeated string vstream_group = 1;
}

message VDevice_configure_Request {
    uint32 handle = 1;
    bytes hef = 2;
    repeated ProtoNamedConfigureNetworkParams configure_params_map = 3;
    uint32 pid = 4;
}

message VDevice_configure_Reply {
    uint32 status = 1;
    repeated uint32 networks_handles = 2;
}

message VDevice_get_physical_devices_ids_Request {
    uint32 handle = 1;
}

message VDevice_get_physical_devices_ids_Reply {
    uint32 status = 1;
    repeated string devices_ids = 2;
}

message VDevice_get_default_streams_interface_Request {
    uint32 handle = 1;
}

message VDevice_get_default_streams_interface_Reply {
    uint32 status = 1;
    uint32 stream_interface = 2;
}

message ConfiguredNetworkGroup_make_input_vstream_params_Request {
    uint32 handle = 1;
    bool quantized = 2;
    uint32 format_type = 3;
    uint32 timeout_ms = 4;
    uint32 queue_size = 5;
    string network_name = 6;
}

message ProtoNamedVStreamParamsMap {
    repeated ProtoNamedVStreamParams vstream_params_map = 1;
}

message ConfiguredNetworkGroup_make_input_vstream_params_Reply {
    uint32 status = 1;
    ProtoNamedVStreamParamsMap vstream_params_map = 2;
}    

message ConfiguredNetworkGroup_make_output_vstream_params_Request {
    uint32 handle = 1;
    bool quantized = 2;
    uint32 format_type = 3;
    uint32 timeout_ms = 4;
    uint32 queue_size = 5;
    string network_name = 6;
}

message ConfiguredNetworkGroup_make_output_vstream_params_Reply {
    uint32 status = 1;
    ProtoNamedVStreamParamsMap vstream_params_map = 2;
}

message ConfiguredNetworkGroup_make_output_vstream_params_groups_Request {
    uint32 handle = 1;
    bool quantized = 2;
    uint32 format_type = 3;
    uint32 timeout_ms = 4;
    uint32 queue_size = 5;
}

message ConfiguredNetworkGroup_make_output_vstream_params_groups_Reply {
    uint32 status = 1;
    repeated ProtoNamedVStreamParamsMap vstream_params_groups = 2;
}

message ConfiguredNetworkGroup_name_Request {
    uint32 handle = 1;
}

message ConfiguredNetworkGroup_name_Reply {
    uint32 status = 1;
    string network_group_name = 2;
}

message ConfiguredNetworkGroup_get_network_infos_Request {
    uint32 handle = 1;
}

message ConfiguredNetworkGroup_get_network_infos_Reply {
    uint32 status = 1;
    repeated string network_infos = 2;
}

message ConfiguredNetworkGroup_get_all_stream_infos_Request {
    uint32 handle = 1;
    string network_name = 2;
}

message ConfiguredNetworkGroup_get_all_stream_infos_Reply {
    uint32 status = 1;
    repeated ProtoStreamInfo stream_infos = 2;
}

message ConfiguredNetworkGroup_get_default_stream_interface_Request {
    uint32 handle = 1;
}

message ConfiguredNetworkGroup_get_default_stream_interface_Reply {
    uint32 status = 1;
    uint32 stream_interface = 2;
}

message ConfiguredNetworkGroup_get_output_vstream_groups_Request {
    uint32 handle = 1;
}

message ConfiguredNetworkGroup_get_output_vstream_groups_Reply {
    uint32 status = 1;
    repeated ProtoVStreamGroup output_vstream_groups = 2;
}

message ConfiguredNetworkGroup_get_vstream_infos_Request {
    uint32 handle = 1;
    string network_name = 2;
}

message ConfiguredNetworkGroup_get_latency_measurement_Request {
    uint32 handle = 1;
    string network_name = 2;
}

message ConfiguredNetworkGroup_get_vstream_infos_Reply {
    uint32 status = 1;
    repeated ProtoVStreamInfo vstream_infos = 2;
}

message ConfiguredNetworkGroup_is_scheduled_Request {
    uint32 handle = 1;
}

message ConfiguredNetworkGroup_is_scheduled_Reply {
    uint32 status = 1;
    bool is_scheduled = 2;
}

message ConfiguredNetworkGroup_set_scheduler_timeout_Request {
    uint32 handle = 1;
    uint32 timeout_ms = 2;
    string network_name = 3;
}

message ConfiguredNetworkGroup_set_scheduler_timeout_Reply {
    uint32 status = 1;
}

message ConfiguredNetworkGroup_set_scheduler_threshold_Request {
    uint32 handle = 1;
    uint32 threshold = 2;
    string network_name = 3;
}

message ConfiguredNetworkGroup_set_scheduler_threshold_Reply {
    uint32 status = 1;
}

message ConfiguredNetworkGroup_set_scheduler_priority_Request {
    uint32 handle = 1;
    uint32 priority = 2;
    string network_name = 3;
}

message ConfiguredNetworkGroup_set_scheduler_priority_Reply {
    uint32 status = 1;
}

message ConfiguredNetworkGroup_get_latency_measurement_Reply {
    uint32 status = 1;
    uint32 avg_hw_latency = 2;
}

message ConfiguredNetworkGroup_is_multi_context_Request {
    uint32 handle = 1;
}

message ConfiguredNetworkGroup_is_multi_context_Reply {
    uint32 status = 1;
    bool is_multi_context = 2;
}

message ConfiguredNetworkGroup_get_config_params_Request {
    uint32 handle = 1;
}

message ConfiguredNetworkGroup_get_config_params_Reply {
    uint32 status = 1;
    ProtoConfigureNetworkParams params = 2;
}

message InputVStream_write_Request {
    uint32 handle = 1;
    bytes data = 2;
}

message InputVStream_write_Reply {
    uint32 status = 1;
}

message OutputVStream_read_Request {
    uint32 handle = 1;
    uint32 size = 2;
}

message OutputVStream_read_Reply {
    uint32 status = 1;
    bytes data = 2;
}

message VStream_get_frame_size_Request {
    uint32 handle = 1;
}

message VStream_get_frame_size_Reply {
    uint32 status = 1;
    uint32 frame_size = 2;
}

message InputVStream_flush_Request {
    uint32 handle = 1;
}

message InputVStream_flush_Reply {
    uint32 status = 1;
}

message VStream_name_Request {
    uint32 handle = 1;
}

message VStream_name_Reply {
    uint32 status = 1;
    string name = 2;
}

message VStream_network_name_Request {
    uint32 handle = 1;
}

message VStream_network_name_Reply {
    uint32 status = 1;
    string network_name = 2;
}

message VStream_abort_Request {
    uint32 handle = 1;
}

message VStream_abort_Reply {
    uint32 status = 1;
}

message VStream_resume_Request {
    uint32 handle = 1;
}

message VStream_resume_Reply {
    uint32 status = 1;
}

message VStream_get_user_buffer_format_Request {
    uint32 handle = 1;
}

message VStream_get_user_buffer_format_Reply {
    uint32 status = 1;
    ProtoHailoFormat user_buffer_format = 2;
}

message VStream_get_info_Request {
    uint32 handle = 1;
}

message VStream_get_info_Reply {
    uint32 status = 1;
    ProtoVStreamInfo vstream_info = 2;
}